consensus/types/src/beacon_state.rs

```diff
@@ -1669,6 +1669,13 @@ impl<T: EthSpec> BeaconState<T> {
         };
         Ok(sync_committee)
     }
+
+    pub fn compute_merkle_proof(&self, depth: usize, index: usize) -> Vec<H256> {
+        let leaves = self.tree_hash_leaves();
+        let tree = merkle_proof::MerkleTree::create(&leaves, depth);
+        let (_, proof) = tree.generate_proof(index, depth);
+        proof
+    }
```

Comment:
Sorry but I don't think this will give the right result. The merkle proof needs to take into account the nested structured of the `BeaconState` and how its fields are packed for hashing. It's not as straight-forward as concatenating the leaves of the `BeaconState`'s fields recursively.

The spec for generalized indices and merkle multi-proofs is here: https://github.com/ethereum/consensus-specs/blob/dev/ssz/merkle-proofs.md. To implement it fully is going to be _hard_ because it requires treating struct/`Container` fields as integers at runtime, and indexing into types using "paths" that are combinations of integer indices and struct field names. I think it will require an extensive redesign of the `tree_hash_derive` macros and probably some extra code generation.

Instead I think we should special-case our implementation for the `BeaconState`. The hashing of the `BeaconState` fields we are interested in is relatively simple.

The leaves of the merkle tree for the `BeaconState` are the `tree_hash_root`s of all its fields in order. For the sync committees, this is all we need and we can construct a proof immediately. For the finalized root I think we just need to patch in the `epoch`. Here's a sketch:

```rust
fn compute_merkle_proof(&self, generalized_index: usize) -> Result<Vec<H256>, Error> {
    // 1. Convert generalized index to field index.
    let field_index = match generalized_index {
        CURRENT_SYNC_COMMITTEE_INDEX | CURRENT_SYNC_COMMITTEE_INDEX => {
            // Sync committees are top-level fields, subtract off the generalized indices
            // for the internal nodes. Result should be 22 or 23, the field offset of the committee
            // in the `BeaconState`:
            // https://github.com/ethereum/consensus-specs/blob/dev/specs/altair/beacon-chain.md#beaconstate
            generalized_index - NUM_BEACON_STATE_HASH_TREE_ROOT_LEAVES
        }
        FINALIZED_ROOT_INDEX => {
            // Finalized root is the right child of `finalized_checkpoint`, divide by two to get
            // the generalized index of `state.finalized_checkpoint`.
            let finalized_checkpoint_generalized_index = generalized_index / 2;
            // Subtract off the internal nodes. Result should be 105/2 - 32 = 20 which matches
            // position of `finalized_checkpoint` in `BeaconState`.
            finalized_checkpoint_generalized_index - NUM_BEACON_STATE_HASH_TREE_ROOT_LEAVES
        }
        _ => return Err(Error::IndexNotSupported(generalized_index)),
    };

    // 2. Get all `BeaconState` leaves.
    // The most efficient way to get these at the moment is from the tree hash cache.
    // We can copy the guts of this function (or generalize it):
    // https://github.com/sigp/lighthouse/blob/6d5a2b509fac7b6ffe693866f58ba49989f946d7/consensus/types/src/beacon_state/tree_hash_cache.rs#L213-L371
    let leaves = ...;

    // 3. Make deposit tree.
    // Use the depth of the `BeaconState` fields (i.e. `log2(32) = 5`).
    let depth = CURRENT_SYNC_COMMITTEE_PROOF_LEN;
    let tree = merkle_proof::MerkleTree::create(&leaves, depth);
    let (_, mut proof) = tree.generate_proof(index, depth);

    // 4. If we're proving the finalized root, patch in the finalized epoch to complete the proof.
    if generalized_index == FINALIZED_ROOT_INDEX {
        proof.push(self.finalized_checkpoint.epoch.tree_hash_root());
    }

    Ok(proof)
}
```

Reply:
Okay, I re-adjusted my code thanks for the guidance, I hope you get better! will look into unit test/integration tests soon!

Reply:
Awesome, looks good!